rosbot_xl_gazebo/test/test_utils.py

```diff
@@ -15,95 +15,80 @@
 
 import rclpy
 from rclpy.node import Node
-from rclpy.executors import SingleThreadedExecutor
 
 from threading import Event
-from threading import Thread
 
 from geometry_msgs.msg import Twist
 from nav_msgs.msg import Odometry
-from sensor_msgs.msg import LaserScan, Image, PointCloud2
+from sensor_msgs.msg import Imu, JointState
 
 
-class SimulationTest:
+class SimulationTestNode(Node):
     __test__ = False
+
     # The inaccuracies in measurement uncertainties and wheel slippage
     # cause the rosbot_xl_base_controller to determine inaccurate odometry.
     ACCURACY = 0.10  # 10% accuracy
+    VELOCITY_STABILIZATION_DELAY = 2
 
     def __init__(self, name="test_node", namespace=None):
-        self.VELOCITY_STABILIZATION_DELAY = 3
-        self.vel_stabilization_time_event = Event()
+        super().__init__(name, namespace=namespace)
+        self.cmd_vel_pub = self.create_publisher(Twist, "cmd_vel", 10)
 
+        # Robot callback
+        self.joint_sub = self.create_subscription(
+            JointState, "joint_states", self.joint_states_callback, 10
+        )
+        self.controller_sub = self.create_subscription(
+            Odometry, "rosbot_xl_base_controller/odom", self.controller_callback, 10
+        )
+        self.ekf_sub = self.create_subscription(
+            Odometry, "odometry/filtered", self.ekf_callback, 10
+        )
+        self.imu_sub = self.create_subscription(Imu, "imu_broadcaster/imu", self.imu_callback, 10)
+
+        # Timer - send cmd_vel and check if the time needed for speed stabilization has elapsed
+        self.timer = self.create_timer(0.1, self.timer_callback)
+
+        # Destination velocity for cmd_vel
         self.v_x = 0.0
         self.v_y = 0.0
         self.v_yaw = 0.0
-        self.twist = None
-
-        self.controller_odom_flag = False
-        self.ekf_odom_flag = False
-        self.scan_event = Event()
-        self.camera_color_event = Event()
-        self.camera_points_event = Event()
-
-        self.ros_context = rclpy.Context()
 
-        self.ros_executor = SingleThreadedExecutor()
-        self.node = Node(name, namespace=namespace)
-        self.ros_executor.add_node(self.node)
+        # Debug values
+        self.controller_twist = None
+        self.ekf_twist = None
+
+        # Robot test flags and event
+        self.is_controller_msg = False
+        self.is_controller_odom_correct = False
+        self.is_ekf_msg = False
+        self.is_ekf_odom_correct = False
+        self.is_imu_msg = False
+        self.is_joint_msg = False
+        self.robot_initialized_event = Event()
+        self.vel_stabilization_time_event = Event()
 
-        # Use simulation time to correct run on slow machine
+        # Parameters Use simulation time to correct run on slow machine
```

Comment:
I think the comment is not right. Using `use_sim_time` parameter changes the  clock source from POSIX time to time published at the `/clock` topic.
```suggestion
        # Using /clock topic as time source
```

Reply:
It's not quite wrong because /clock is publish from gazebo, so if gazebo run slowly the simulation time can also decrease

Reply:
Therefore the slow run of simulation is not a purpose to use `use_sim_time` parameter. The real reason is using time from simulation is to work with the time  from the beginning of the simulation engine not a POSIX time.